### PR TITLE
Add pyright to test suite, pyrightconfig.json

### LINKED ISSUE #5048: Add pyright to tests
I propose to add pyright to our CI test suite. pyright is a stable type checker and probably the fastest to implement new typing features. Also:

* pyright is frequently used as it is the standard type checker for the popular Visual Studio Code IDE/editor.
* pyright has many useful warnings, not found in other type checkers.
* Not running pyright as part of our CI complicates the job for pyright's maintainers as evidenced by the frequent clean-up PRs by Eric Traut. This will be even more of an issue when pyright decides not to ship third-party stubs.

I think the discussion in #5035 shows that adding pyright to CI is uncontroversial. Cc'ing @JelleZijlstra @hauntsaninja and @erictraut.

### PR DESCRIPTION
**Pull request by jakebailey**
Fixes #5048.

Building on #5051, with some modifications.

### CODE REVIEW

**Comment by srittau** [comment]
I have merged the warning fixes to master now.

**Comment by jakebailey** [comment]
Still a couple of errors, it seems.

```
/home/runner/work/typeshed/typeshed/stdlib/fractions.pyi
  123:9 - error: Overload 1 for "__pow__" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
/home/runner/work/typeshed/typeshed/stdlib/functools.pyi
  94:9 - error: Overload 1 for "register" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
  113:13 - error: Overload 1 for "register" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
/home/runner/work/typeshed/typeshed/stdlib/tempfile.pyi
  329:5 - error: Overload 1 for "mkdtemp" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
/home/runner/work/typeshed/typeshed/stdlib/asyncio/tasks.pyi
  54:5 - error: Overload 1 for "gather" overlaps overload 7 and returns an incompatible type (reportGeneralTypeIssues)
  58:5 - error: Overload 2 for "gather" overlaps overload 8 and returns an incompatible type (reportGeneralTypeIssues)
  66:5 - error: Overload 3 for "gather" overlaps overload 9 and returns an incompatible type (reportGeneralTypeIssues)
  75:5 - error: Overload 4 for "gather" overlaps overload 10 and returns an incompatible type (reportGeneralTypeIssues)
  85:5 - error: Overload 5 for "gather" overlaps overload 11 and returns an incompatible type (reportGeneralTypeIssues)
/home/runner/work/typeshed/typeshed/stubs/pkg_resources/pkg_resources/__init__.pyi
  57:9 - error: Overload 1 for "best_match" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
  61:9 - error: Overload 1 for "obtain" overlaps overload 2 and returns an incompatible type (reportGeneralTypeIssues)
11 errors, 0 warnings, 0 infos
```

**Comment by erictraut** [comment]
Oh, I moved those errors to a new diagnostic rule. I'll need to publish a new version of pyright (1.1.114) before you can complete this. I'll do this in the next few hours.

**Comment by jakebailey** [comment]
Pyright is now passing. The only things that are missing are changes to allow externally setting the Python version and `sys.platform` so that a matrix test can be run like mypy.

**Review by srittau — APPROVED** [review]
🎉

**Comment by hauntsaninja** [comment]
Thanks all for getting this in place! Would be nice to get to not having any stdlib stubs in the exclude list.

**Comment by erictraut** [comment]
Yeah, I agree. "tkinter" is going to require a lot of work. The others less so.

Some of the third-party stubs in the exclude list are also very close to passing and could get there with relatively little effort.

File: pyrightconfig.json [hunk 1]
```diff
@@ -0,0 +1,95 @@
+{
+    "typeshedPath": ".",
+    "include": [
+        "stdlib",
+        "stubs"
+    ],
+    "exclude": [
+        "**/@python2",
+        "stdlib/crypto/__init__.pyi",
```
Comment (jakebailey) [inline pyrightconfig.json:9]: This file doesn't seem to exist, now that I look.

Comment (erictraut) [reply pyrightconfig.json:9]: Yeah, you're right. You can delete that line or leave it in and we'll delete it the next time we update it. It was left over from an earlier version of this config.

Comment (jakebailey) [reply pyrightconfig.json]: Deleted it now.

**Pull request merged**